asyncpg/connection.py

```diff
@@ -296,17 +296,23 @@ def is_in_transaction(self):
             stmt_name = ''
 
         statement = await self._protocol.prepare(stmt_name, query, timeout)
-        ready = statement._init_types()
-        if ready is not True:
-            types, intro_stmt = await self.__execute(
-                self._intro_query, (list(ready),), 0, timeout)
-            self._protocol.get_settings().register_data_types(types)
+        need_reprepare = False
+        types_with_missing_codecs = statement._init_types()
+        while types_with_missing_codecs:
```

Comment:
Need a MAX_ITER constant and raise an exception if we loop more times than it (in case we have a bug in out logic that will make it easier for people to report it)